File: c/src/classical/substitution/fractionated/trifid.c
/**
 * @file trifid.c
 * @brief Trifid Cipher implementation with custom alphabet support
 */

#include "cryptology/classical/substitution/fractionated/trifid.h"
#include "cryptology/classical/substitution/polygraphic/alphabet_utils.h"
#include <string.h>
#include <ctype.h>
#include <stdbool.h>
#include <math.h>
#include <stdlib.h>

#define MAX_KEY_SIZE 256
#define CUBE_SIZE 3
#define DEFAULT_ALPHABET "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

static int find_char_in_cube(char cube[CUBE_SIZE][CUBE_SIZE][CUBE_SIZE], char c, int *layer, int *row, int *col) {
    for (int l = 0; l < CUBE_SIZE; l++) {
        for (int r = 0; r < CUBE_SIZE; r++) {
            for (int c_idx = 0; c_idx < CUBE_SIZE; c_idx++) {
                if (cube[l][r][c_idx] == c) {
                    *layer = l;
                    *row = r;
                    *col = c_idx;
                    return 0;
                }
            }
        }
    }
    return -1;
}

static int create_trifid_cube(const char *key, const char *alphabet, char cube[CUBE_SIZE][CUBE_SIZE][CUBE_SIZE]) {
    if (!key || !alphabet) {
        return -1;
    }
    
    // Handle custom alphabets
    char processed_alphabet[MAX_KEY_SIZE];
    if (strcmp(alphabet, DEFAULT_ALPHABET) != 0) {
        // Combine similar letters for non-English alphabets
        if (combine_similar_letters(alphabet, "auto", processed_alphabet, sizeof(processed_alphabet)) != 0) {
            return -1;
        }
        
        // For Trifid, we need exactly 27 characters (3x3x3)
        if (strlen(processed_alphabet) > 27) {
            processed_alphabet[27] = '\0';  // Truncate to 27
        } else if (strlen(processed_alphabet) < 27) {
            // Pad with X
            int len = strlen(processed_alphabet);
            for (int i = len; i < 27; i++) {
                processed_alphabet[i] = 'X';
            }
            processed_alphabet[27] = '\0';
        }
    } else {
        // Standard English alphabet (I and J combined, 26 letters + 1 padding)
        strcpy(processed_alphabet, "ABCDEFGHIKLMNOPQRSTUVWXYZ");  // 25 letters
        strcat(processed_alphabet, "X");  // Pad to 26
        strcat(processed_alphabet, "X");  // Pad to 27
    }
    
    // Remove duplicates while preserving order, convert to uppercase
    char key_clean[MAX_KEY_SIZE];
    bool seen[256] = {false};
    int pos = 0;
    
    for (size_t i = 0; key[i] != '\0' && pos < MAX_KEY_SIZE - 1; i++) {
        char c = toupper((unsigned char)key[i]);
        if (c >= 'A' && c <= 'Z' && !seen[(unsigned char)c]) {
            key_clean[pos++] = c;
            seen[(unsigned char)c] = true;
        }
    }
    key_clean[pos] = '\0';
    
    // Add remaining letters from processed alphabet
    for (size_t i = 0; processed_alphabet[i] != '\0' && pos < MAX_KEY_SIZE - 1; i++) {
        char c = toupper((unsigned char)processed_alphabet[i]);
        if (!seen[(unsigned char)c]) {
            key_clean[pos++] = c;
            seen[(unsigned char)c] = true;
        }
    }
    key_clean[pos] = '\0';
    
    // Ensure we have exactly 27 characters
    if (strlen(key_clean) > 27) {
        key_clean[27] = '\0';
    } else if (strlen(key_clean) < 27) {
        // Pad with X
        int len = strlen(key_clean);
        for (int i = len; i < 27; i++) {
            key_clean[i] = 'X';
        }
        key_clean[27] = '\0';
    }
    
    // Create 3x3x3 cube
    for (int layer = 0; layer < CUBE_SIZE; layer++) {
        for (int row = 0; row < CUBE_SIZE; row++) {
            for (int col = 0; col < CUBE_SIZE; col++) {
                int index = layer * 9 + row * 3 + col;
                if (index < (int)strlen(key_clean)) {
                    cube[layer][row][col] = key_clean[index];
                } else {
                    cube[layer][row][col] = 'X';  // Padding
                }
            }
        }
    }
    
    return 0;
}

static int prepare_text(const char *input, const char *alphabet, char *output, size_t output_size) {
    if (!input || !output || output_size == 0) {
        return -1;
    }
    
    size_t input_len = strlen(input);
    size_t output_pos = 0;
    
    for (size_t i = 0; i < input_len && output_pos < output_size - 1; i++) {
        char c = toupper((unsigned char)input[i]);
        if (c >= 'A' && c <= 'Z') {
            // Handle custom alphabets
            if (strcmp(alphabet, DEFAULT_ALPHABET) != 0) {
                // Apply language-specific replacements
                const char *language = detect_language(alphabet);
                if (strcmp(language, "turkish") == 0) {
                    // Turkish character replacements
                    if (c == 'Ç') c = 'C';
                    else if (c == 'Ğ') c = 'G';
                    else if (c == 'I') c = 'I';
                    else if (c == 'Ö') c = 'O';
                    else if (c == 'Ş') c = 'S';
                    else if (c == 'Ü') c = 'U';
                }
            } else {
                // Standard English: Replace J with I
                if (c == 'J') {
                    c = 'I';
                }
            }
            output[output_pos++] = c;
        }
    }
    
    output[output_pos] = '\0';
    return 0;
}

int trifid_encrypt(const char *plaintext, const char *key, 
                   char *result, size_t result_size) {
    return trifid_encrypt_with_alphabet(plaintext, key, DEFAULT_ALPHABET, result, result_size);
}

int trifid_decrypt(const char *ciphertext, const char *key,
                  char *result, size_t result_size) {
    return trifid_decrypt_with_alphabet(ciphertext, key, DEFAULT_ALPHABET, result, result_size);
}

int trifid_encrypt_with_alphabet(const char *plaintext, const char *key, const char *alphabet,
                                 char *result, size_t result_size) {
    if (!plaintext || !key || !result || result_size == 0) {
        return -1;
    }
    
    char cube[CUBE_SIZE][CUBE_SIZE][CUBE_SIZE];
    
    if (create_trifid_cube(key, alphabet, cube) != 0) {
        return -1;
    }
    
    char prepared_text[MAX_KEY_SIZE];
    if (prepare_text(plaintext, alphabet, prepared_text, sizeof(prepared_text)) != 0) {
        return -1;
    }
    
    size_t prepared_len = strlen(prepared_text);
    if (prepared_len == 0) {
        result[0] = '\0';
        return 0;
    }
    
    // Convert each letter to coordinates
    int *layers = malloc(prepared_len * sizeof(int));
    int *rows = malloc(prepared_len * sizeof(int));
    int *cols = malloc(prepared_len * sizeof(int));
    
    if (!layers || !rows || !cols) {
        free(layers);
        free(rows);
        free(cols);
        return -1;
    }
    
    int coord_count = 0;
    for (size_t i = 0; i < prepared_len; i++) {
        int layer, row, col;
        if (find_char_in_cube(cube, prepared_text[i], &layer, &row, &col) == 0) {
            layers[coord_count] = layer;
            rows[coord_count] = row;
            cols[coord_count] = col;
            coord_count++;
        }
    }
    
    if (coord_count == 0) {
        free(layers);
        free(rows);
        free(cols);
        result[0] = '\0';
        return 0;
    }
    
    // Fractionation: write all layers, then all rows, then all columns
    int *fractionated = malloc(coord_count * 3 * sizeof(int));
    if (!fractionated) {
        free(layers);
        free(rows);
        free(cols);
        return -1;
    }
    
    for (int i = 0; i < coord_count; i++) {
        fractionated[i] = layers[i];
        fractionated[coord_count + i] = rows[i];
        fractionated[coord_count * 2 + i] = cols[i];
    }
    
    // Read triplets of coordinates to get new letters
    size_t result_pos = 0;
    for (int i = 0; i < coord_count * 3 && result_pos < result_size - 1; i += 3) {
        if (i + 2 < coord_count * 3) {
            int layer = fractionated[i];
            int row = fractionated[i + 1];
            int col = fractionated[i + 2];
            if (layer < CUBE_SIZE && row < CUBE_SIZE && col < CUBE_SIZE) {
                result[result_pos++] = cube[layer][row][col];
            }
        }
    }
    
    result[result_pos] = '\0';
    
    free(layers);
    free(rows);
    free(cols);
    free(fractionated);
    
    return 0;
}

int trifid_decrypt_with_alphabet(const char *ciphertext, const char *key, const char *alphabet,
                                 char *result, size_t result_size) {
    if (!ciphertext || !key || !result || result_size == 0) {
        return -1;
    }
    
    char cube[CUBE_SIZE][CUBE_SIZE][CUBE_SIZE];
    
    if (create_trifid_cube(key, alphabet, cube) != 0) {
        return -1;
    }
    
    char prepared_text[MAX_KEY_SIZE];
    if (prepare_text(ciphertext, alphabet, prepared_text, sizeof(prepared_text)) != 0) {
        return -1;
    }
    
    size_t prepared_len = strlen(prepared_text);
    if (prepared_len == 0) {
        result[0] = '\0';
        return 0;
    }
    
    // Convert each letter to coordinates
    int *coords = malloc(prepared_len * 3 * sizeof(int));
    if (!coords) {
        return -1;
    }
    
    int coord_count = 0;
    for (size_t i = 0; i < prepared_len; i++) {
        int layer, row, col;
        if (find_char_in_cube(cube, prepared_text[i], &layer, &row, &col) == 0) {
            coords[coord_count * 3] = layer;
            coords[coord_count * 3 + 1] = row;
            coords[coord_count * 3 + 2] = col;
            coord_count++;
        }
    }
    
    if (coord_count == 0) {
        free(coords);
        result[0] = '\0';
        return 0;
    }
    
    // Defractionation: separate layers, rows, and columns
    int *layers = malloc(coord_count * sizeof(int));
    int *rows = malloc(coord_count * sizeof(int));
    int *cols = malloc(coord_count * sizeof(int));
    
    if (!layers || !rows || !cols) {
        free(coords);
        free(layers);
        free(rows);
        free(cols);
        return -1;
    }
    
    for (int i = 0; i < coord_count; i++) {
        layers[i] = coords[i * 3];
        rows[i] = coords[i * 3 + 1];
        cols[i] = coords[i * 3 + 2];
    }
    
    // Interleave layers, rows, and columns
    size_t result_pos = 0;
    for (int i = 0; i < coord_count && result_pos < result_size - 1; i++) {
        int layer = layers[i];
        int row = rows[i];
        int col = cols[i];
        if (layer < CUBE_SIZE && row < CUBE_SIZE && col < CUBE_SIZE) {
            result[result_pos++] = cube[layer][row][col];
        }
    }
    
    result[result_pos] = '\0';
    
    free(coords);
    free(layers);
    free(rows);
    free(cols);
    
    return 0;
}

File: python/cryptology/classical/substitution/fractionated/trifid.py
"""
Trifid Cipher implementation.

The Trifid cipher is a fractionated substitution cipher that uses a 3x3x3 cube.
It works by:
1. Converting each letter to its layer, row, and column coordinates
2. Writing all layer coordinates, then all row coordinates, then all column coordinates
3. Reading triplets of coordinates to get new letters

This 3D fractionation technique provides even more security than Bifid.
"""

from typing import List, Tuple, Optional, Dict, Any
import cryptology.alphabets as ALPHABETS
from ..polygraphic.alphabet_utils import combine_similar_letters, get_square_size, create_square_alphabet
from ..polygraphic.monoalphabetic_squares import create_monoalphabetic_square

DEFAULT_ALPHABET = ALPHABETS.ENGLISH_ALPHABET


def _create_trifid_cube(
    key: str, 
    alphabet: str = DEFAULT_ALPHABET,
    square_type: str = "standard",
    mono_params: Optional[Dict[str, Any]] = None
) -> List[List[List[str]]]:
    """
    Create a Trifid cube from the key and alphabet.
    
    Args:
        key: The keyword to use for the cube
        alphabet: The alphabet to use (default: English)
        square_type: Type of square ("standard", "caesar", "atbash", "affine", "keyword")
        mono_params: Parameters for monoalphabetic-based squares
        
    Returns:
        A cube as a list of lists of lists
    """
    # Determine target cube size (3x3x3 or 3x3x4) based on alphabet length
    def compute_dims(target_len: int) -> Tuple[int, int, int]:
        # Layers and rows fixed at 3; columns vary
        layers = 3
        rows = 3
        cols = (target_len + (layers * rows) - 1) // (layers * rows)
        # Clamp to at least 3
        if cols < 3:
            cols = 3
        return layers, rows, cols

    # Handle monoalphabetic-based square generation
    if square_type in ["caesar", "atbash", "affine", "keyword"]:
        # Use shared monoalphabetic square generation
        if mono_params is None:
            mono_params = {}
            if square_type == "caesar":
                mono_params["shift"] = 3  # Default Caesar shift
            elif square_type == "affine":
                mono_params = {"a": 5, "b": 8}  # Default Affine params
        
        # Create square string (can be 5x5 or 6x6)
        square_string = create_monoalphabetic_square(square_type, alphabet, mono_params)
        
        # Extract characters for Trifid cube
        # Convert multiline string to single line
        square_chars = "".join(square_string.split()).lower()
        # Decide target length: 27 or 36
        target_len = 27 if len(square_chars) <= 27 else 36
        square_chars = square_chars[:target_len].ljust(target_len, 'x')
        L, R, C = compute_dims(target_len)
        # Create cube from transformed alphabet
        cube = []
        index = 0
        for layer in range(L):
            cube_layer = []
            for row in range(R):
                cube_row = []
                for col in range(C):
                    if index < len(square_chars):
                        cube_row.append(square_chars[index])
                    else:
                        cube_row.append('x')
                    index += 1
                cube_layer.append(cube_row)
            cube.append(cube_layer)
        return cube
    
    # Handle custom alphabets (standard cube creation)
    if alphabet != DEFAULT_ALPHABET:
        # Preserve all characters for non-English alphabets
        processed_alphabet = alphabet
        # Choose 27 or 36 target size
        target_len = 27 if len(processed_alphabet) <= 27 else 36
        if len(processed_alphabet) > target_len:
            processed_alphabet = processed_alphabet[:target_len]
        elif len(processed_alphabet) < target_len:
            processed_alphabet = processed_alphabet.ljust(target_len, 'x')
    else:
        # Standard English alphabet (I and J combined, 26 letters + 1 padding)
        processed_alphabet = "abcdefghiklmnopqrstuvwxyz"  # 25 letters (no j)
        processed_alphabet = processed_alphabet.ljust(27, 'x')  # Pad to 27
    
    # Remove duplicates while preserving order, keep lowercase
    key_clean = ""
    seen = set()
    
    for char in key.lower():
        if char.isalpha() and char not in seen:
            key_clean += char
            seen.add(char)
    
    # Add remaining letters from processed alphabet
    for char in processed_alphabet.lower():
        if char not in seen:
            key_clean += char
            seen.add(char)
    
    # Ensure we have target characters (27/36)
    target_len = 27 if len(processed_alphabet) <= 27 else 36
    if len(key_clean) > target_len:
        key_clean = key_clean[:target_len]
    elif len(key_clean) < target_len:
        key_clean = key_clean.ljust(target_len, 'x')
    
    # Create cube with computed dims
    L, R, C = compute_dims(target_len)
    cube = []
    index = 0
    for layer in range(L):
        cube_layer = []
        for row in range(R):
            cube_row = []
            for col in range(C):
                if index < len(key_clean):
                    cube_row.append(key_clean[index])
                else:
                    cube_row.append('x')
                index += 1
            cube_layer.append(cube_row)
        cube.append(cube_layer)
    
    return cube


def _find_position(cube: List[List[List[str]]], char: str) -> Tuple[int, int, int]:
    """
    Find the layer, row, and column position of a character in the cube.
    
    Args:
        cube: The Trifid cube
        char: The character to find
        
    Returns:
        A tuple of (layer, row, column) coordinates
    """
    char = char.lower()
    
    L = len(cube)
    R = len(cube[0])
    C = len(cube[0][0])
    for layer in range(L):
        for row in range(R):
            for col in range(C):
                if cube[layer][row][col] == char:
                    return layer, row, col
    
    raise ValueError(f"Character '{char}' not found in cube")


def _prepare_text(text: str, alphabet: str = DEFAULT_ALPHABET) -> str:
    """
    Prepare text for encryption by cleaning and handling special cases.
    
    Args:
        text: The input text
        alphabet: The alphabet being used
        
    Returns:
        Cleaned text ready for encryption
    """
    # Clean the text
    text_clean = ""
    for char in text.lower():
        # Keep letters and digits for ciphertext robustness
        if char.isalpha() or char.isdigit():
            # English: replace j with i on letters only
            if alphabet == DEFAULT_ALPHABET and char == 'j':
                char = 'i'
            text_clean += char
    
    return text_clean


def encrypt(
    plaintext: str, 
    key: str, 
    alphabet: str = DEFAULT_ALPHABET,
    square_type: str = "standard",
    mono_params: Optional[Dict[str, Any]] = None
) -> str:
    """
    Encrypt plaintext using the Trifid cipher.
    
    Args:
        plaintext: The text to encrypt
        key: The keyword for the Trifid cube
        alphabet: The alphabet to use (default: English)
        
    Returns:
        The encrypted text
    """
    if not plaintext or not key:
        return ""
    
    # Create the Trifid cube
    cube = _create_trifid_cube(key, alphabet, square_type, mono_params)
    
    # Prepare the text
    text_clean = _prepare_text(plaintext, alphabet)
    
    if not text_clean:
        return ""
    
    
    # Block-wise processing with period
    period = 5
    L = len(cube)
    R = len(cube[0])
    C = len(cube[0][0])
    
    result = ""
    for start in range(0, len(text_clean), period):
        block = text_clean[start:start + period]
        # Convert each letter to coordinates
        layers: List[int] = []
        rows: List[int] = []
        cols: List[int] = []
        for char in block:
            try:
                layer, row, col = _find_position(cube, char)
                layers.append(layer)
                rows.append(row)
                cols.append(col)
            except ValueError:
                continue
        if not layers:
            continue
        # Fractionation: write all layers, then rows, then cols (within block)
        # Produce ciphertext by pairing corresponding indices
        for i in range(len(layers)):
            layer = layers[i]
            row = rows[i]
            col = cols[i]
            if layer < L and row < R and col < C:
                result += cube[layer][row][col]
    return result


def decrypt(
    ciphertext: str, 
    key: str, 
    alphabet: str = DEFAULT_ALPHABET,
    square_type: str = "standard",
    mono_params: Optional[Dict[str, Any]] = None
) -> str:
    """
    Decrypt ciphertext using the Trifid cipher.
    
    Args:
        ciphertext: The text to decrypt
        key: The keyword for the Trifid cube
        alphabet: The alphabet to use (default: English)
        
    Returns:
        The decrypted text
    """
    if not ciphertext or not key:
        return ""
    
    # Create the Trifid cube
    cube = _create_trifid_cube(key, alphabet, square_type, mono_params)
    
    # Prepare the text
    text_clean = _prepare_text(ciphertext, alphabet)
    
    if not text_clean:
        return ""
    
    
    # Block-wise processing with period
    period = 5
    L = len(cube)
    R = len(cube[0])
    C = len(cube[0][0])
    
    result = ""
    for start in range(0, len(text_clean), period):
        block = text_clean[start:start + period]
        # Convert each letter to coordinates for the block
        coords: List[Tuple[int, int, int]] = []
        for char in block:
            try:
                layer, row, col = _find_position(cube, char)
                coords.append((layer, row, col))
            except ValueError:
                continue
        if not coords:
            continue
        # Defractionation for the block: rebuild original index-wise
        layers_digits = [c[0] for c in coords]
        rows_digits = [c[1] for c in coords]
        cols_digits = [c[2] for c in coords]
        for i in range(len(coords)):
            layer = layers_digits[i]
            row = rows_digits[i]
            col = cols_digits[i]
            if layer < L and row < R and col < C:
                result += cube[layer][row][col]
    return result
